tests/hypothesis_strategies/kernels.py

```diff
@@ -12,3 +12,16 @@
     lengthscale_prior=st.one_of(st.none(), priors),
 )
 """A strategy that generates Matern kernels."""
+
+base_kernels = st.one_of([matern_kernels])
+"""A strategy that generates base kernels to be used within more complex kernels."""
+
+scale_kernels = st.builds(
+    ScaleKernel,
+    base_kernel=base_kernels,
+    outputscale_prior=st.one_of(st.none(), priors),
+)
+"""A strategy that generates Scale kernels."""
+
+kernels = st.one_of([base_kernels, scale_kernels])
```

Comment:
To me, this is not a semantically consistent grouping of kernels (i.e. a `ScaleKernel` contains a `MaternKernel` and is therefore a higher-order object in a sense). 

Instead of having a strategy that creates either of the two (and then later additional base kernels such as RBF....) I would have rather expected that a strategy decides for each of the base kernels randomly whether or not to wrap them in a scale kernel. Would make more sense to me...

Reply:
this quagmire would be solved if we wouldnt have a `ScaleKernel` class at all and instead provide baybe kernels with an `outputscale_prior` optional arg similar to `lenghscale_prior` dealing with the gpytorch scalekernel internally instead

Reply:
I disagree here: A `ScaleKernel` is kernel that users can provide in the exact same way they can provide a `MaternKernel`. Hence, the list `kernels` which should just contain all kernels that users could provide contain both. And they are not a higher-order object, as they are both just inherited from the same base class.
This is why there is the explicit strategy `base_kerneles`- this is where stuff where additional strategies for RBFkernels and so on would go, so this call would actually not change until we have stuff like "AdditiveKernels"

Reply:
after the meeting I agree with you, this is simply a 50/50 switch between base kernels with/without outer scaling kernel
from a wording perspective it might be suboptimal though, what we call `scale_kernels` does not deliver scale kernels but base_kernels wrapped in scale kernels, but I have no good suggestions for something else

Reply:
We could do something similar to gpytorch, they call such kernels "Decoration/Composition kernels". This would then yield three lists (`base_kernels`, `decoration_kernels`, `composition_kernels`), and one could (if it makes sense at some point) even extend the kernel class hierarchy by using this classification (although not necessary atm)

Reply:
Wait wait, you both are now two steps ahead (and I think in the wrong direction, tbh). But first things first:
- @AVHopp: IMO, a scale kernel absolutely _is_ at a higher level than a Matern kernel – you even rank it as such yourself by calling the other kernels "base"-kernels for exactly that reason. Just because something is on the same level of some arbitrary class hierarchy doesn't mean that things are on the same level semantically. Arbitrary example:
  ```python
  class Thing:
      pass
  
  class Wall(Thing):
     pass
  
  class House(Thing):
     pass
  ```
  Following that logic,  you would then need to conclude that houses and walls are "same level", even though houses are made of walls? Same with the kernels. You simply can't create a `ScaleKernel` without a regular one, but you can create the regular one without a scale – that's I think quite literally my definition of a "higher-level" object. This holds true in gpytorch but also on a mathematical level: you can't apply a scaler to something if there is no "something"
- Second aspect: think in terms of the shrinking mechanism in hypothesis: If you want to shrink a kernel "removing" the scale is one of the things you would want to do because it gives you a simpler object (same reason as above: "kernel + scale" is by definition more complex than just "kernel")
- So I agree that having several strategies for differnt parts of the functionality is fine, e.g. `base_kernel` and `kernels` where the latter contains all sorts of stuff. But in terms of shrinkage, it does not make any sense to me to have `kernel A` hard-coded in the same strategy as `Scale(Kernel A)`... rather, the strategy should cleverly compose the latter in a shrinkage-compatible way. Just like you would not provide a manual list of all possible sum or product kernels. Instead, we need a mechanism that emits combinations of increasing complexity automatically and can shrink them back if necessary.

Reply:
seems switching around the two list entries is already fulfilling the shrinking idea? Why does it make no sense, the shrinkage will go for the non-scaled version just as you suggest?

and I wonder, do we always have to adhere to having the `simplest` item int he beginning of the list, or does it also make sense to have the `more defaulty` item in the beginning of the lists?

Reply:
Using the right order in the list is OK, but **only** for the very simple case of scale vs no scale (i.e. one characteristic that we need to change). But the list-like approach no longer works for multiple such characteristics. The problem is that you try to artificially imprint a one-dimensional structure on something that is inherently multi-dimensional, meaning you not only have to choose a "random" order/projection but it become increasingly cumbersome to do so with more and more decision variables. And in fact, even in the current setting, you'd be ignoring stuff that SHOULD be part of the strategy, namely the lenghtscale and output scale priors....

So how would your approach look like if you consider these as well, plus additive/multiplicative kernels? Would you then say: first base kernels, then base kernels with scale, then base kernels with scale + scale prior, then with lengthscale prior, then both priors, then sum kernels of size two, then with where one of them has scale, .... You get what I mean, right?

Why even try to manually manage that when hypothesis is specifically designed to handle this? Instead of flattening into a list, let hypothesis flexibly compose the kernels, where each decision is orthogonal to the others and shrinking works out of the box.

Reply:
Note: Composite strategy with base kernel as argument and bool whether or not to scale

Reply:
Attempted to at least improve the hypothesis strategy :D Please have a look and either resolve or direct me in a more suitable direction regarding the proper use of hypothesis :)

Reply:
What about the `ScaleKernel` built as `composite` ?

Reply:
Will change (and resolve this mega discussion here for the sake of creating a smaller one at the corresponding point in the code.)